decidim-proposals/app/commands/decidim/proposals/endorse_proposal.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+module Decidim
+  module Proposals
+    # A command with all the business logic when a user endorses a proposal.
+    class EndorseProposal < Rectify::Command
+      # Public: Initializes the command.
+      #
+      # proposal     - A Decidim::Proposals::Proposal object.
+      # current_user - The current user.
+      # current_group_id- (optional) The current_grup that is endorsing the Proposal.
+      def initialize(proposal, current_user, current_group_id = nil)
+        @proposal = proposal
+        @current_user = current_user
+        @current_group_id = current_group_id
+      end
+
+      # Executes the command. Broadcasts these events:
+      #
+      # - :ok when everything is valid, together with the proposal vote.
+      # - :invalid if the form wasn't valid and we couldn't proceed.
+      #
+      # Returns nothing.
+      def call
+        build_proposal_endorsement
+        return broadcast(:invalid) unless @endorsement.valid?
+        @endorsement.save!
+        broadcast(:ok, endorsement)
+      end
+
+      attr_reader :endorsement
+
+      private
+
+      def build_proposal_endorsement
```

Comment:
We're sacrificing explicitness for safety which I'm not sure its better.

Also, as another alternative, I could get rid of the attribute and make build_proposal_endorsement return the endorsement and use it as a local variable (this would be better than my original implementation and keep the code more debuggable if typos).

```
def call
    endorsement = build_proposal_endorsement
    endorsement.save ? broadcast(:ok, endorsement) : broadcast(:invalid)
end
```
what do you think?

Reply:
Works for me! 😄